Fix gh-6397 isSafeFile fails on deeply nested $_FILES

nikosdion opened the pull request:
This PR fixes gh-6397 "JFilterInput::isSafeFile() fail on very deep $_FILES array"
# Testing instructions
- Download and install the Joomla! 3 sample component from https://github.com/joomla/Joomla-3.2-Hello-World-Component/archive/step-10-adding-decorations-to-the-backend.zip
- Open the file administrator/components/com_helloworld/models/forms/helloworld.xml and add the following before the closing `</fieldset>`

``` xml
        <field type="file" name="document"  label="Document" />
        <fields name="files">
            <fields name="files0">
                <field type="file" name="file"  label="File 1" />
            </fields>
            <fields name="files1">
                <field type="file" name="file"  label="File 2" />
            </fields>
            <fields name="files2">
                <field type="file" name="file"  label="File 3" />
            </fields>
        </fields>
        <fields name="images">
            <fields name="images0">
                <field type="file" name="file"  label="Images 1" />
            </fields>
            <fields name="images1">
                <field type="file" name="file"  label="Images 2" />
            </fields>
        </fields>
```
- Open the file administrator/components/com_helloworld/controllers/helloworld.php and add the following between the two curly braces

``` php
    public function save($key = null, $urlVar = null)
    {
        $app = JFactory::getApplication();
        $files = $app->input->files->get('jform');

        die('CHECK');

        return parent::save($key, $urlVar); // TODO: Change the autogenerated stub
    }
```
- Go to Global Configuration and set Error Reporting to Development
- Go to Components, Hello World and click on the "Hello, world" link to edit the item.
- Click each file selection field and choose a small image from your computer. It's best to select files where each one is less than 100Kb to prevent upload issues with PHP itself.
- Click on the Save & Close button

**Result before the patch**
A lot of PHP warnings

**Result after the patch**
No PHP warnings, just the word "CHECK" printed on your page
# Technical notes

There were two corrective actions taken:
- The safe file check was moved into get(). This may have a performance impact on unoptimised code calling $app->input->files->get() repeatedly over the same files, but there's no work around due to the way PHP reports uploaded files. Sorry :(
- The JFilterInput::isSafeFile method was modified to handle nested file arrays, either in the raw $_FILES format or as nested arrays of file descriptors (these are the only two possible variations that exist)

Signed-off-by: Nicholas K. Dionysopoulos nicholas@akeebabackup.com

Commit:
Fix gh-6397 isSafeFile fails on deeply nested $_FILES

This PR fixes gh-6397 "JFilterInput::isSafeFile() fail on very deep $_FILES array"

# Testing instructions

* Download and install the Joomla! 3 sample component from https://github.com/joomla/Joomla-3.2-Hello-World-Component/archive/step-10-adding-decorations-to-the-backend.zip
* Open the file administrator/components/com_helloworld/models/forms/helloworld.xml and add the following before the closing `</fieldset>`

```xml
        <field type="file" name="document"  label="Document" />
        <fields name="files">
            <fields name="files0">
                <field type="file" name="file"  label="File 1" />
            </fields>
            <fields name="files1">
                <field type="file" name="file"  label="File 2" />
            </fields>
            <fields name="files2">
                <field type="file" name="file"  label="File 3" />
            </fields>
        </fields>
        <fields name="images">
            <fields name="images0">
                <field type="file" name="file"  label="Images 1" />
            </fields>
            <fields name="images1">
                <field type="file" name="file"  label="Images 2" />
            </fields>
        </fields>
```

* Open the file administrator/components/com_helloworld/controllers/helloworld.php and add the following between the two curly braces

```php
	public function save($key = null, $urlVar = null)
	{
		$app = JFactory::getApplication();
		$files = $app->input->files->get('jform');

		die('CHECK');

		return parent::save($key, $urlVar); // TODO: Change the autogenerated stub
	}
```

* Go to Global Configuration and set Error Reporting to Development
* Go to Components, Hello World and click on the "Hello, world" link to edit the item.
* Click each file selection field and choose a small image from your computer. It's best to select files where each one is less than 100Kb to prevent upload issues with PHP itself.
* Click on the Save & Close button

**Result before the patch**
A lot of PHP warnings

**Result after the patch**
No PHP warnings, just the word "CHECK" printed on your page

# Technical notes

There were two corrective actions taken:

* The safe file check was moved into get(). This may have a performance impact on unoptimised code calling $app->input->files->get() repeatedly over the same files, but there's no work around due to the way PHP reports uploaded files. Sorry :(

* The JFilterInput::isSafeFile method was modified to handle nested file arrays, either in the raw $_FILES format or as nested arrays of file descriptors (these are the only two possible variations that exist)

Signed-off-by: Nicholas K. Dionysopoulos <nicholas@akeebabackup.com>

Fedik commented:
thanks @nikosdion 

**test**
work good

brianteeman commented:
@nikosdion travis isnt happy<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6454">issues.joomla.org/joomla-cms/6454</a>.</sub>

Commit:
Make Travis happy

Nevermind that the code in question was copied from JInputFiles as it is currently present in Joomla! 3.4 and you have told phpcs to ignore it because it doesn't conform to the coding standards...

nikosdion commented:
It's not really my fault. Travis chokes on a line I copied verbatim from JInputFiles. Apparently core code doesn't conform to the core coding standards... Anyway, I updated the PR to make Travis a happy chap.

brianteeman commented:
I only said travis wasnt happy. I did not apportion blame :)

nikosdion commented:
I know, Brian. Let me and Michael complain about the state of the core :)

wilsonge commented on libraries/joomla/filter/input.php:
Can we realign this please?

Commit:
Docblock alignment

nikosdion commented:
@wilsonge I fixed the alignment

Kubik-Rubik commented:
@cheesegrits We need your feedback ASAP. Thanks!
